### ISSUE TITLE
Do not emit unused code with Request Delegate Generator

### ISSUE THREAD

**Issue by martincostello** [dotnet/aspnetcore#48381]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

While looking at the code for `RequestDelegateGeneratorSources` to see if I could see the source of the bug for #48379 I noticed that private static constants are emitted for all HTTP methods, whether or not they are actually used:

https://github.com/dotnet/aspnetcore/blob/39564d529f84f7a3bbac5b28ba11060e8ac30375/src/Http/Http.Extensions/gen/RequestDelegateGeneratorSources.cs#L648-L652

In the following generated code, only `GetVerb` and `PostVerb` are used:

```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.AspNetCore.Http.RequestDelegateGenerator, Version=8.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60", "8.0.0.0")]
internal static class GenerateRouteBuilderEndpoints
{
    private static readonly string[] GetVerb = new[] { global::Microsoft.AspNetCore.Http.HttpMethods.Get };
    private static readonly string[] PostVerb = new[] { global::Microsoft.AspNetCore.Http.HttpMethods.Post };
    private static readonly string[] PutVerb = new[]  { global::Microsoft.AspNetCore.Http.HttpMethods.Put };
    private static readonly string[] DeleteVerb = new[] { global::Microsoft.AspNetCore.Http.HttpMethods.Delete };
    private static readonly string[] PatchVerb = new[] { global::Microsoft.AspNetCore.Http.HttpMethods.Patch };

    internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapGet(
        this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
        [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
        global::System.Func<global::System.Collections.Generic.IEnumerable<global::MartinCostello.AdventOfCode.PuzzleAttribute>, global::Microsoft.AspNetCore.Http.IResult> handler,
        [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
        [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
    {
        return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(
            endpoints,
            pattern,
            handler,
            GetVerb,
            filePath,
            lineNumber);
    }
    internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapPost(
        this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
        [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
        global::System.Func<global::System.Int32, global::System.Int32, global::Microsoft.AspNetCore.Http.HttpRequest, global::Microsoft.AspNetCore.Http.IFormFile?, global::MartinCostello.AdventOfCode.PuzzleFactory, global::Microsoft.Extensions.Logging.ILogger<global::MartinCostello.AdventOfCode.Puzzle>, global::System.Threading.CancellationToken, global::System.Threading.Tasks.Task<global::Microsoft.AspNetCore.Http.IResult>> handler,
        [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
        [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
    {
        return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(
            endpoints,
            pattern,
            handler,
            PostVerb,
            filePath,
            lineNumber);
    }
    internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder MapGet(
        this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
        [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
        global::System.Func<global::Microsoft.AspNetCore.Http.IResult> handler,
        [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
        [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
    {
        return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(
            endpoints,
            pattern,
            handler,
            GetVerb,
            filePath,
            lineNumber);
    }
}
```

The memory and IL size would be _ever so slightly_ smaller if the fields weren't emitted (assuming the compiler isn't removing them anyway). This is of course a micro-optimisation, so feel free to close this suggestion if the complexity of the generator code isn't worth changing it to detect which of the arrays are actually needed.

### Describe the solution you'd like

The source generator only emits the string arrays that are needed by the endpoints used in the compilation.

**Comment by mitchdenny** [comment]
It is a complexity trade off but we've definitely got logic already where we omit certain types entirely based on the endpoints they are defined.

**Comment by danmoseley** [comment]
@mitchdenny is this the sort of issue where we'd take a community PR? I'm hoping we can apply the 'help wanted' label whenever we'd take a community PR and believe we would have time to review and accept one (which of course is not always the case, eg., if something would likely need a lot of collaboration)

**Comment by mitchdenny** [comment]
I think we would absolutely take a PR on this.

**Comment by martincostello** [comment]
I'll see if I can come up with a PR to address this that doesn't make things too complicated.

**Comment by martincostello** [comment]
Proposed fix in #48426.
